src/test/codegen/sanitizer-cfi-emit-type-metadata-id-itanium-cxx-abi.rs

```diff
@@ -0,0 +1,704 @@
+// Verifies that type metadata identifiers for functions are emitted correctly.
+//
+// ignore-windows
+// needs-sanitizer-cfi
+// compile-flags: -Clto -Cno-prepopulate-passes -Zsanitizer=cfi
+
+#![crate_type="lib"]
+#![feature(adt_const_params, extern_types, inline_const, slice_pattern, type_alias_impl_trait)]
+
+extern crate core;
+use core::ffi::c_void;
+use std::marker::PhantomData;
+
+// User-defined type (structure)
+pub struct Struct1<T> {
+    member1: T,
+}
+
+// User-defined type (enum)
+pub enum Enum1<T> {
+    Variant1(T),
+}
+
+// User-defined type (union)
+pub union Union1<T> {
+    member1: std::mem::ManuallyDrop<T>,
+}
+
+extern {
+    // Extern type
+    type Type1;
+
+    // Extern function
+    fn fn1();
+}
+
+// impl Trait type aliases for helping with defining other complex types (see below)
+pub type Type2 = impl Send;
+pub type Type3 = impl Send;
+pub type Type4 = impl Send;
+pub type Type5 = impl Send;
+pub type Type6 = impl Send;
+pub type Type7 = impl Send;
+pub type Type8 = impl Send;
+pub type Type9 = impl Send;
+pub type Type10 = impl Send;
+pub type Type11 = impl Send;
+pub type Type12 = impl Send;
+
+pub fn fn2<'a>() {
+    // Disambiguated closures
+    let closure1 = || { };
+    let closure2 = || { };
+    let closure3 = || { };
+
+    let _: Type2 = closure1;
+    let _: Type3 = closure2;
+    let _: Type4 = closure3;
+
+    // Type in impl path (i.e., inherent method)
+    let _: Type5 = <[i32]>::len;
+
+    // Type in impl path (i.e., trait method)
+    let _: Type6 = <[i32] as core::slice::SlicePattern>::as_slice;
+
+    // Type in extern path
+    let _: Type7 = fn1;
+
+    // Type in closure path
+    || {
+        pub struct Foo;
+        let _: Type8 = Foo;
+    };
+
+    // Constructor
+    pub struct Bar(i32);
+    let _: Type9 = Bar;
+
+    // Type in const path
+    const {
+        pub struct Baz;
+        fn fn3() -> Type10 { Baz }
+    };
+
+    // Const generics
+    pub struct Qux<T, const N: usize>([T; N]);
+    let _: Type11 = Qux([0; 32]);
+
+    // Lifetimes/regions
+    pub struct Quux<'a>(&'a i32);
+    pub struct Quuux<'a, 'b>(&'a i32, &'b Quux<'b>);
+    let _: Type12 = Quuux;
```

Comment:
Do you have a better example to test lifetimes/regions here following this same model/pattern?

Reply:
Moved comment to up-to-date changes.